src/EnergyPlus/Vectorize.hh

```diff
@@ -0,0 +1,34 @@
+#ifndef EnergyPlus_vectorize_hh_INCLUDED
+#define EnergyPlus_vectorize_hh_INCLUDED
+
+// EnergyPlus Vectorization Support
+
+// ObjexxFCL Headers
+#include <ObjexxFCL/vectorize.hh> // ASSUME_ALIGNED and RESTRICT macros
+
+// C++ Headers
+#include <cassert>
+#include <cstdint>
+
+// EnergyPlus headers
+#include "EnergyPlus.hh"
+
+namespace EnergyPlus {
+
+	int const VEC_LENGTH = VEC_ALIGN / sizeof(Real64);
+
+#define ROUND_TO_VEC_LENGTH(I) (((I) + ((VEC_LENGTH) - 1)) & ~((VEC_LENGTH) - 1))
+
+
+#define EXPLICIT_VECTORIZATION
+
+	// EXPLICIT_VECTORIZATION assumes that vectors are aligned,
+	// which is not a valid assumption on 32-bit builds
+#if defined(_WIN32) && !defined(_WIN64)
```

Comment:
@mbadams5, it's not luck. It's that 64-bit memory allocation libraries
align to 16-bytes, whereas 32-bit libraries align to 8-bytes. Not sure why
this is. Maybe @deadparrot knows. For what it's worth, the solution we have
right now will not scale to AVX because I don't think any library aligns to
32-bytes.

Re: movapd vs. movupd (i.e., _mm_load_pd and _mm_loadu_pd). When I did my
deep dives on this (8 or 9 year ago), this was a pretty big difference. And
when I looked at it here it was still 1-2% overall (not in the loop or in
CIRE, in EnergyPlus). So I decided to keep _mm_load_pd for now because we
can use it on 64-bit builds. I don't think we want to sacrifice performance
on 64-bit builds just to get performance on 32-bit builds which are on the
way out.


On Sat, Dec 12, 2015 at 10:16 PM, Mark Adams <notifications@github.com>
wrote:

> In src/EnergyPlus/Vectorize.hh
> <https://github.com/NREL/EnergyPlus/pull/5278#discussion_r47439715>:
>
> > +
> > +// EnergyPlus headers
> > +#include "EnergyPlus.hh"
> > +
> > +namespace EnergyPlus {
> > +
> > +	int const VEC_LENGTH = VEC_ALIGN / sizeof(Real64);
> > +
> > +#define ROUND_TO_VEC_LENGTH(I) (((I) + ((VEC_LENGTH) - 1)) & ~((VEC_LENGTH) - 1))
> > +
> > +
> > +#define EXPLICIT_VECTORIZATION
> > +
> > +	// EXPLICIT_VECTORIZATION assumes that vectors are aligned,
> > +	// which is not a valid assumption on 32-bit builds
> > +#if defined(_WIN32) && !defined(_WIN64)
>
> @amirroth <https://github.com/amirroth> Why is it not a valid assumption
> for 32-bit builds, but it is for 64-bit builds? Are we just getting lucky
> in 64-bit builds with the (much) larger memory address space compared to
> 32-bit?
>
> I know this is a (slight) performance hit, but why not use _mm_loadu_pd
> <https://software.intel.com/sites/landingpage/IntrinsicsGuide/#techs=SSE,SSE2&text=_mm_loadu_pd&expand=3106,3106>?
> This allows you to load unaligned data from memory. From my research, it
> seems like modern processors have a much smaller performance hit using this
> function versus _mm_load_pd than they used to.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NREL/EnergyPlus/pull/5278/files#r47439715>.
>